spec/models/memorial_spec.rb

```diff
@@ -0,0 +1,182 @@
+require 'rails_helper'
+
+RSpec.describe Memorial, type: :model do
+  describe "validations" do
+    context "validate name" do
+      it {should validate_presence_of(:name)}
+      it {should validate_length_of(:name).is_at_most(Settings.length.digit_50)}
+    end
+
+    context "validate relationship" do
+      it {should validate_presence_of(:relationship)}
+    end
+
+    context "validate biography" do
+      it {should validate_length_of(:biography)
+        .is_at_most(Settings.length.digit_1000)}
+    end
+  end
+
+  describe "enum" do
+    it {should define_enum_for(:privacy_type).with_values(public: 0, private: 1, shared: 2)
+          .with_prefix(true)}
+  end
+
+  describe "associations" do
+    it {should have_many(:placetimes).dependent(:destroy)}
+
+    it {should have_many(:contributions).dependent(:destroy)}
+
+    it {should have_many(:user_relations)
+      .class_name(AccessPrivacy.name).dependent(:destroy)}
+
+    it {should have_many(:shared_users)
+      .through(:user_relations).source(:user)}
+
+    it {should have_one_attached(:avatar)}
+  end
+
+  let!(:owner){create :user}
```

Comment:
Trả lời câu hỏi: Phân biệt let và let!

Reply:
let: dùng để define 1  memoized helper method, nó sẽ không tự động thực thi mà chỉ được chạy khi gọi, giá trị của nó sẽ được lưu trong bộ nhớ đệm ứng với từng khối it (example). Nghĩa là trong 1 khối it nếu nó được gọi nhiều lần, thì nó chỉ thực thi 1 lần, các lần sau thì sẽ lấy lại giá trị lưu trong bộ nhớ.
Sự khác nhau giữa let! và let là let! sẽ được gọi trước khối it nên giá trị của nó đã được lưu lại, nên khi gọi trong it thì sẽ không thực thi nữa.

Reply:
let: dùng để define 1  memoized helper method, nó sẽ không tự động thực thi mà chỉ được chạy khi gọi, giá trị của nó sẽ được lưu trong bộ nhớ đệm ứng với từng khối it (example). Nghĩa là trong 1 khối it nếu nó được gọi nhiều lần, thì nó chỉ thực thi 1 lần, các lần sau thì sẽ lấy lại giá trị lưu trong bộ nhớ.


Sự khác nhau giữa let! và let là let! sẽ được gọi trước khối it nên giá trị của nó đã được lưu lại, nên khi gọi trong it thì sẽ không thực thi nữa.